### ISSUE TITLE
Question Regarding Endpoints

### ISSUE THREAD

**Issue by neilhosey** [IdentityServer/IdentityServer3#2872]
Hi Brock,

We are using IdentityServer3 and its working great.  The business wish to put a REST API in front of each Endpoint.    This obviously gives us a layer of abstraction.

 I am weary of doing this at this point as I am afraid we will break some conditions specified in OAuth 2.0 or OpenID Connect, and as we introduce OAuth2 (and possibly OICD) in the future, we will have to refactor the API contract to match some criteria we missed.  This will be costly as we expose our IdSrc3 to external clients (thru client credential flow).

So basically im wondering would you recommend putting an API in front of IdSrv3, or access it directly? 

Thanks again

Neil

**Comment by leastprivilege** [comment]
Our endpoints are the protocol implementations - i don't see any benefit of abstracting that.

**Comment by lukos** [comment]
What would be the purpose of the REST API? 

idsrv already uses API controllers for some of the endpoints but, importantly, there needs to be user interaction for authentication so having REST APIs, which are generally designed to return json or other simpler data, are not really compatible with the concept of OAuth2 or OIC.

**Comment by neilhosey** [comment]
Yes, I agree I see no benefit of abstraction either, although there was a choice made for a previous implementation of an STS, and I wanted to make sure this wasn't a general practice with IdSrc3 as it doesn't seem right to me.

The only benefit I see is the layer of abstraction for cross cutting concerns, but i believe these can be addressed with a proxy.

Thanks for the input!
